migrations/013.do.meridian-rounds.sql

```diff
@@ -0,0 +1,11 @@
+CREATE TABLE meridian_rounds (
+  -- 0x of f4 address of the MERidian smart contract emitting "round started" events
+  -- We are using TEXT instead of BYTEA for simplicity
+  contract_address TEXT NOT NULL PRIMARY KEY,
```

Comment:
it looks like this table is storing contract version rather than rounds. With meridian rounds I'd expect the primary key to be a composite of contract and round offset. Like this we will only store one row per contract (right?)

Reply:
I'm somewhat confused about this logic (and therefore round-tracker). Could you please give me a rough overview of the algorithm, so that I can continue with the review?

Reply:
Maybe it's better to start the review in the function using this data: https://github.com/filecoin-station/spark-api/pull/61/files#diff-16e2f7b1cd688374798168c1ef69985dbd460a6a451142a27b29c5d8c348312dR22

What's the best place to write down documentation for this logic? Should I create a new ADR?

---

There are three cases I want to handle:

1. Business as usual - the IE contract advanced the round by one
2. Fresh start, e.g. a new spark-api instance is deployed, or we deploy this PR to an existing instance.
3. Upgrade of the IE contract

For each IE version (defined as the smart contract address), I am keeping track of three fields:
- `contractAddress` 
- `sparkRoundOffset`
- `lastSparkRound`

Whenever a new IE round is started, we know the current IE round number (`meridianRoundNumber`)

Let me explain how are the different cases handled.

**Business as usual**

We want to map IE round number to SPARK round number. This assumes we have already initialised our DB for the current IE contract version we are working with.

```
sparkRound = meridianRoundNumber + sparkRoundOffset
```

For example, if we deploy my PR and then observe IE round 123, then `sparkRoundOffset` is `-122` and we calculate the spark round as
```
123 + (-122) = 1
```

We update the record for the current IE contract address to set `last_spark_round = sparkRound`.

**Fresh start**

There is no record in our DB. We want to map the current IE round number to SPARK round 1. Also, we want to setup `sparkRoundOffset` so that the algorithm above produces correct SPARK round numbers.

```
sparkRound = 1
sparkRoundOffset = sparkRound - meridianRoundNumber
```

We insert a new record to our DB with the address of the current IE contract, `sparkRoundOffset`, and `last_spark_round = sparkRound`.

**Upgrading IE contract**

We have one or more existing records in our DB. We know what is the last SPARK round that we calculated from the previous version of the IE contract (`lastSparkRound`). We also know what is the round number of the new IE contract.

```
sparkRound = lastSparkRound + 1
sparkRoundOffset = sparkRound - meridianRoundNumber
```

We insert a new record to our DB with the address of the current IE contract, `sparkRoundOffset`, and `last_spark_round = sparkRound`.

If you are wondering how to find out what is the last SPARK round that we calculated from the previous version of the IE contract - we can easily find it in our DB:

```sql
SELECT last_spark_round FROM meridian_rounds ORDER BY last_spark_round DESC LIMIT 1
```

Reply:
I captured the explanation in a code comment.